### PR TITLE
Use correct CoreLib configuration for shared framework linker trimming

### PR DESCRIPTION
**Pull request by layomia**
From https://github.com/dotnet/runtime/pull/45821#issuecomment-741858345:

`GetTargetPath` is returning the path to CoreLib based on the libraries configuration (`-lc`), rather than the runtime configuration (`-rc`). This means that when we do a build such as `build -s clr+libs -rc Release`, we are passing `Debug` libraries and `Debug` CoreLib, rather than `Debug` libraries and `Release` CoreLib to the linker which is not expected.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
Curious, why does the `GetTargetPath` target not work for you? What are you trying to solve with this PR?

**Comment by layomia** [comment]
@ViktorHofer - `GetTargetPath` is returning the path to CoreLib based on the libraries configuration (`-lc`), rather than the runtime configuration (`-rc`). This means that when we do a build such as `build -s clr+libs -rc Release`, we are passing `Debug` libraries and `Debug` CoreLib, rather than `Debug` libraries and `Release` CoreLib which is not expected.

Any thoughts on how we could flow the `RuntimeConfiguration` prop to `GetTargetPath`, or thoughts on the current approach using `CoreCLRArtifactsPath`?

cc @eerhardt, @joperezr

**Comment by ViktorHofer** [comment]
Yes, you could something similar to what we do for the ProjectReference to CoreLib: https://github.com/dotnet/runtime/blob/dcbf6e9dee81434e5be937897c8445df4e71c70e/src/libraries/Directory.Build.targets#L175-L179

**Comment by ViktorHofer** [comment]
You can pass in the "Properties" and "RemoveProperties" properties to the MSBuild task: https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-task?view=vs-2019

Pass in `$(RuntimeConfiguration)` as the value of configuration in the `Properties` property and just to be on the safe side, remove the `Platform` and `TargetFramework` properties.

**Comment by ViktorHofer** [comment]
This is a point in time problem until we start binplacing CoreLib...

**Comment by layomia** [comment]
> This is a point in time problem until we start binplacing CoreLib...

Yup, https://github.com/dotnet/runtime/pull/40172#discussion_r465090284.

**Comment by layomia** [comment]
> You can pass in the "Properties" and "RemoveProperties" properties to the MSBuild task: https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-task?view=vs-2019
>
> Pass in $(RuntimeConfiguration) as the value of configuration in the Properties property and just to be on the safe side, remove the Platform and TargetFramework properties.

Thanks, trying this out now. What problem do we avoid by removing the `Platform` and `TargetFramework` properties?

**Comment by ViktorHofer** [comment]
> What problem do we avoid by removing the Platform and TargetFramework properties?

We do this for performance as msbuild caches project evaluations based on the global properties passed in. Doing this here might reduce an unnecessary project evaluation of CoreLib by using the same global properties as for the ProjectReference to CoreLib.

File: src/libraries/illink-sharedframework.targets [hunk 1]
```diff
@@ -30,8 +30,10 @@
       <ILLinkArgs>$(ILLinkArgs) --nowarn $(LinkerNoWarn)</ILLinkArgs>
     </PropertyGroup>
 
-     <!-- Retrieve CoreLib's targetpath via GetTargetPath as it isn't binplaced yet. -->
+    <!-- Retrieve CoreLib's path as it isn't binplaced alongside the libraries. -->
     <MSBuild Projects="$(CoreLibProject)"
+             Properties="Configuration=$(RuntimeConfiguration)"
```
Comment (ViktorHofer) [inline src/libraries/illink-sharedframework.targets:35]: With this, the Configuration property is always passed in as a global property even if the RuntimeConfiguration already matches the current (libraries) configuration. We should keep this in sync with how we set global properties for the ProjectReference to CoreLib.

Comment (ViktorHofer) [reply src/libraries/illink-sharedframework.targets:35]:
```
         <Properties Condition="'$(RuntimeFlavor)' == 'CoreCLR' and 
                                '$(Configuration)' != '$(CoreCLRConfiguration)'">Configuration=$(CoreCLRConfiguration)</Properties> 
         <Properties Condition="'$(RuntimeFlavor)' == 'Mono' and 
                                '$(Configuration)' != '$(MonoConfiguration)'">Configuration=$(MonoConfiguration)</Properties> 
```

**Comment by ericstj** [comment]
@layomia @eerhardt @joperezr is this PR still relevant?

**Comment by layomia** [comment]
@ericsj - I believe so. IIRC @eerhardt had a work around locally for the [issue](https://github.com/dotnet/runtime/pull/45821#issuecomment-741858345), but we need a proper fix checked in for correctness. I'll go ahead and drive this in.

**Comment by eerhardt** [comment]
> IIRC @eerhardt had a work around locally for the issue

Yes, my workaround was to build a Debug CoreLib.

This is what gets passed to the linker today with `dotnet build .\src\libraries\src.proj /t:ILLinkTrimSharedFramework /p:BuildingNetCoreAppVertical=true /p:RuntimeConfiguration=Release`. Note the CoreLib path is to `Debug` even though I said `/p:RuntimeConfiguration=Release`.
image

**Comment by ericstj** [comment]
I see, I wasn't sure if we had passed this point in time :)

> This is a point in time problem until we start binplacing CoreLib...

Can you either close this PR and file an issue, or address @ViktorHofer's feedback and merge?

**Comment by layomia** [comment]
I'll address the feedback and merge, since it's a small effort at this point.

File: src/libraries/illink-sharedframework.targets [hunk 2]
```diff
@@ -30,8 +30,20 @@
...
-     <!-- Retrieve CoreLib's targetpath via GetTargetPath as it isn't binplaced yet. -->
-    <MSBuild Projects="$(CoreLibProject)"
+    <!-- Retrieve CoreLib's path as it isn't binplaced alongside the libraries. -->
+    <ItemGroup>
+      <CoreLibProjectWithRuntimeConfig Include="$(CoreLibProject)">
+        <!-- Don't flow TargetFramework and Platform to use same inputs and outputs as the CoreLib's build as part of the runtime. -->
+        <UndefineProperties>$(UndefineProperties);TargetFramework;Platform</UndefineProperties>
+        <!-- If conflicting, manually set the Configuration property of the CoreLib project so that it aligns with the specified RuntimeConfiguration in the libraries' build. -->
+        <Properties Condition="'$(RuntimeFlavor)' == 'CoreCLR' and
+                               '$(Configuration)' != '$(CoreCLRConfiguration)'">Configuration=$(CoreCLRConfiguration)</Properties>
```
Comment (eerhardt) [inline src/libraries/illink-sharedframework.targets:40]: Just a question, mostly for @ViktorHofer. Why do we have both `$(CoreCLRConfiguration)` and `$(MonoConfiguration)`? Why not just have `$(RuntimeConfiguration)`? Do we expect to be able to build the repo using different configurations for coreclr and mono?

File: src/libraries/illink-sharedframework.targets [hunk 3]
```diff
@@ -30,8 +30,20 @@
       <ILLinkArgs>$(ILLinkArgs) --nowarn $(LinkerNoWarn)</ILLinkArgs>
     </PropertyGroup>
 
-     <!-- Retrieve CoreLib's targetpath via GetTargetPath as it isn't binplaced yet. -->
-    <MSBuild Projects="$(CoreLibProject)"
+    <!-- Retrieve CoreLib's path as it isn't binplaced alongside the libraries. -->
```
Comment (layomia) [inline src/libraries/illink-sharedframework.targets:33]: @ViktorHofer - is there a tracking issue for this work?

File: src/libraries/illink-sharedframework.targets (hunk 2 above)
Comment (ViktorHofer) [reply src/libraries/illink-sharedframework.targets:40]: There is a `RuntimeConfiguration` property but in this case we can't use that as we need to respect the inferred `CoreCLRConfiguration` and `MonoConfiguration` properties.

> Do we expect to be able to build the repo using different configurations for coreclr and mono?

Yes

File: src/libraries/illink-sharedframework.targets (hunk 3 above)
Comment (ViktorHofer) [reply src/libraries/illink-sharedframework.targets:33]: https://github.com/dotnet/runtime/issues/43095

**Review by ViktorHofer — APPROVED** [review]

**Review by ViktorHofer — APPROVED** [review]

**Comment by layomia** [comment]
CI failures are unrelated - dup of https://github.com/dotnet/runtime/issues/47374.

**Pull request merged**